internal/manifests/targetallocator/targetallocator.go

```diff
@@ -0,0 +1,43 @@
+// Copyright The OpenTelemetry Authors
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package targetallocator
+
+import (
+	"sigs.k8s.io/controller-runtime/pkg/client"
+
+	"github.com/open-telemetry/opentelemetry-operator/internal/reconcileutil"
+)
+
+func Build(params reconcileutil.Params) ([]client.Object, error) {
+	var manifests []client.Object
+	if !params.Instance.Spec.TargetAllocator.Enabled {
+		return nil, nil
+	}
+	objects := []reconcileutil.ObjectCreator{
+		reconcileutil.Creator(ConfigMap),
+		reconcileutil.CreatorWithoutError(Deployment),
+		reconcileutil.CreatorWithoutError(ServiceAccount),
+		reconcileutil.CreatorWithoutError(Service),
+	}
+	for _, object := range objects {
+		res, err := object(params.Config, params.Log, params.Instance)
+		if err != nil {
+			return nil, err
+		} else if res != nil && res.DeepCopyObject() != nil {
```

Comment:
honestly this really confused me, but my understanding is that because res is an interface, res should never be nil because it's going to be `client.Object(nil)` when what we want to ask is if it is actually nil which is why we do the deepcopy. Open to other ideas here.

Reply:
https://stackoverflow.com/questions/13476349/check-for-nil-and-nil-interface-in-go for more info

Reply:
Ah, it's actually a check for "typed nil"? I'd leave a comment explaining this here then, as it's not very obvious.

Reply:
can do! Do you think i should just use reflect here? Or is the deep copy sufficient?

Reply:
I'm not a Go expert by any measure, but here's how I see the possibilities:

1. Stay with the current behaviour. Most of the time we're going to be doing unnecessary deep copies, which is far from ideal.
2. Use reflect to check if the value is a nil pointer. Effective, but a bit hacky.
3. Only check for plain nil, and allow the output to contain typed nil values. The callers downstream who will actually cast this to a concrete type (`*v1.Deployment` for example) can check for the remaining nil safely on their own.
4. Maybe there's an interface in apimachinery we could cast this to that would let us do a cheap check?

From reading the linked mailing list thread, I think 3 is supposed to be the correct solution to this problem, and personally I'm fine with it. The additional check downstream is realistically something we need to do anyway.